[FIX] account: Fix tax recomputation with tax cash rounding

dutrieuc opened the pull request:
Description of the issue/feature this PR addresses:

Step to reproduce:

1. Set up Cash rounding for Tax: Accounting> Configuration > cash rounding
2. Create cash rounding with precision 0.05, rounding strategy to "modify tax amount"
3. Create an invoice - add a line with price of 45 (to have the rounding) and 7.7% tax
4. In the invoice > "Other info" tab > add the rounding method
6. Change the quantity or price

Current behavior before PR:
tax doesn't update if there is a rounding line
Desired behavior after PR is merged:
tax should be updated.

Why:
What happens is that Odoo is trying to update the new tax amount on the wrong line (the rounded tax amount line).

That's because the `existing_after` dict is poorly defined. It is defined by the `existing()` inner function of `_sync_dynamic_line` thanks to the "tax_key" of each line.​

The two account move lines (VAT line and rounding on the VAT line) share the same tax_key,​ and because of that the VAT line is overwritten by the rounding line in the dict returned by`existing()`.

Then the cash rounding is recomputed from the VAT line(that still holds the outdated values) and overwrites the updated values on the rounding line.

Solution:
Adding the type of line in the tax_key to differentiate between a VAT line and a rounding VAT line


opw-3224743

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/114974).

aliyatastemirova reviewed (commented):
Thanks for the fix.
A few small things: 
The commit message is missing space after [FIX], which makes it fail style tests.
The ticket id convention is writing it with a dash: ~closes 3224743~ opw-3224743

Also, I think it would be very nice to add a test for this case :slightly_smiling_face:

dutrieuc commented:
Here you go @yosa-odoo @aliyatastemirova

aliyatastemirova commented on addons/account/tests/test_account_move_line_tax_details.py:
```diff
@@ -654,6 +654,47 @@ def test_round_globally_rounding(self):
+            [
+                {
+                    'base_line_id': base_lines[0].id,
+                    'tax_line_id': tax_lines.id,
```
I think it might be better to also specify `tax_lines[0].id` or something? As `tax_lines` can theoretically have multiple records (not in this test, but generally)

qdp-odoo commented:
hello @dutrieuc , thanks for having a look at this issue.

However, I'm very confused by your issue description: 
1) your use case contains steps that don't look necessary (using a recurent product, creating a SO...)
2) it fails to provide an important piece of configuration: how the cash rounding must be configured (create new line or change tax amount? at the end you talk about the "tax rounding line", so I suppose it's the later but the problem also occurs with the other option, did you notice?? )
3) the test case you added doesn't reflect the said use case. In one you modify the quantity, in the second the price_unit
4) finally, yes, I'm able to reproduce the problem but your fix doesn't seem to do anything to help...

So maybe it's just me being too confused by the commit message, but I definitely need you to rewrite it and prove to me it effectively solves the issue. :nerd_face: 

thanks :pray:

dutrieuc commented:
hey @qdp-odoo , good points, thanks for looking into it.

>     1. your use case contains steps that don't look necessary (using a recurent product, creating a SO...)
That's how we first reproduced the issue but writing the tests I removed some parts that didn't look needed to reproduce. I updated the PR description to remove the useless parts and add details, I'll update the commit one as well if we can agree on how to reproduce :sweat_smile: 

>     2. it fails to provide an important piece of configuration: how the cash rounding must be configured (create new line or change tax amount? at the end you talk about the "tax rounding line", so I suppose it's the later but the problem also occurs with the other option, did you notice?? )
Hm no I didn't. On my particular issue the issue probably wasn't present with global rounding as the total didn't need any rounding.

>     3. the test case you added doesn't reflect the said use case. In one you modify the quantity, in the second the price_unit
I think it doesn't matter, what matters is changing the subtotal for the line.

>     4. finally, yes, I'm able to reproduce the problem but your fix doesn't seem to do anything to help...
 That test is failing when I remove the fix (and my manual reproduction as well) so I think it is solving something.

How did you reproduce ? I guess there is a lot of ways there can be dupplicate keys and I only patched one.

william-andre commented on addons/account/models/account_move_line.py:
```diff
@@ -915,6 +915,7 @@ def _compute_tax_key(self):
                     'tax_tag_ids': [(6, 0, line.tax_tag_ids.ids)],
                     'partner_id': line.partner_id.id,
                     'move_id': line.move_id.id,
+                    'display_type': line.display_type,
```
If you are adding it in here, you must also add it in the keys of `compute_all_tax`, otherwise it will burn through ids instead of recycling them (look at how `to_delete` and `to_create` are merged as a `write` in `_sync_dynamic_line`)

dutrieuc replied:
@william-andre  I followed your advice and that lead me to patch a test as well as it caused one of the invoice lines to split into two lines. I think it should be equivalent as it balances to the same amounts but let me know if you think that's an issue.

<details>
  <summary>The details for the two new lines if needed</summary>
[{'id': 138, 'analytic_distribution': False, 'analytic_precision': 0, 'move_id': (45, 'Draft Invoice INV/2019/00001'), 'journal_id': (240, 'Customer Invoices'), 'company_id': (35, 'company_1_data'), 'company_currency_id': (2, 'USD'), 'move_name': 'INV/2019/00001', 'parent_state': 'draft', 'date': datetime.date(2019, 1, 1), 'ref': False, 'is_storno': False, 'sequence': 10000, 'move_type': 'out_invoice', 'account_id': (1572, '251000 Tax Received'), 'name': 'Tax 15%', 'debit': 0.0, 'credit': 150.0, 'balance': -150.0, 'cumulated_balance': 0.0, 'currency_rate': 1.0, 'amount_currency': -150.0, 'currency_id': (2, 'USD'), 'is_same_currency': True, 'partner_id': (124, 'partner_a'), 'reconcile_model_id': False, 'payment_id': False, 'statement_line_id': False, 'statement_id': False, 'tax_ids': [], 'group_tax_id': False, 'tax_line_id': (168, 'Tax 15%'), 'tax_group_id': (2, 'Tax 15%'), 'tax_base_amount': 1000.0, 'tax_repartition_line_id': (702, 'account.tax.repartition.line,702'), 'tax_tag_ids': [], 'tax_audit': '', 'tax_tag_invert': True, 'amount_residual': 0.0, 'amount_residual_currency': 0.0, 'reconciled': False, 'full_reconcile_id': False, 'matched_debit_ids': [], 'matched_credit_ids': [], 'matching_number': False, 'is_account_reconcile': False, 'account_type': 'liability_current', 'account_internal_group': 'liability', 'account_root_id': (50053, '25'), 'display_type': 'tax', 'product_id': False, 'product_uom_id': False, 'product_uom_category_id': False, 'quantity': 0.0, 'date_maturity': False, 'price_unit': 0.0, 'price_subtotal': 0.0, 'price_total': 0.0, 'discount': 0.0, 'term_key': False, 'tax_key': {'tax_repartition_line_id': 702, 'group_tax_id': False, 'account_id': 1572, 'currency_id': 2, 'analytic_distribution': False, 'tax_ids': [(6, 0, [])], 'tax_tag_ids': [(6, 0, [])], 'partner_id': 124, 'move_id': 45, 'display_type': 'tax'}, 'compute_all_tax': {}, 'compute_all_tax_dirty': False, 'epd_key': False, 'epd_needed': False, 'epd_dirty': True, 'analytic_line_ids': [], 'discount_date': False, 'discount_amount_currency': 0.0, 'discount_balance': 0.0, 'discount_percentage': 0.0, 'blocked': False, 'is_refund': False, '__last_update': datetime.datetime(2023, 4, 5, 8, 52, 6, 842107), 'display_name': 'INV/2019/00001 Tax 15%', 'create_uid': (24, 'Because I am accountman!'), 'create_date': datetime.datetime(2023, 4, 5, 8, 52, 6, 842107), 'write_uid': (24, 'Because I am accountman!'), 'write_date': datetime.datetime(2023, 4, 5, 8, 52, 6, 842107), 'move_attachment_ids': [], 'expected_pay_date': False, 'is_downpayment': False, 'sale_line_ids': [], 'asset_ids': [], 'non_deductible_tax_value': 0.0, 'followup_line_id': False, 'last_followup_date': False, 'next_action_date': False, 'invoice_date': datetime.date(2019, 1, 1), 'invoice_origin': False}]

---

[{'id': 140, 'analytic_distribution': False, 'analytic_precision': 0, 'move_id': (45, 'Draft Invoice INV/2019/00001'), 'journal_id': (240, 'Customer Invoices'), 'company_id': (35, 'company_1_data'), 'company_currency_id': (2, 'USD'), 'move_name': 'INV/2019/00001', 'parent_state': 'draft', 'date': datetime.date(2019, 1, 1), 'ref': False, 'is_storno': False, 'sequence': 10000, 'move_type': 'out_invoice', 'account_id': (1572, '251000 Tax Received'), 'name': 'Tax 15%', 'debit': 11.25, 'credit': 0.0, 'balance': 11.25, 'cumulated_balance': 0.0, 'currency_rate': 1.0, 'amount_currency': 11.25, 'currency_id': (2, 'USD'), 'is_same_currency': True, 'partner_id': (124, 'partner_a'), 'reconcile_model_id': False, 'payment_id': False, 'statement_line_id': False, 'statement_id': False, 'tax_ids': [], 'group_tax_id': False, 'tax_line_id': (168, 'Tax 15%'), 'tax_group_id': (2, 'Tax 15%'), 'tax_base_amount': -75.0, 'tax_repartition_line_id': (702, 'account.tax.repartition.line,702'), 'tax_tag_ids': [], 'tax_audit': '', 'tax_tag_invert': True, 'amount_residual': 0.0, 'amount_residual_currency': 0.0, 'reconciled': False, 'full_reconcile_id': False, 'matched_debit_ids': [], 'matched_credit_ids': [], 'matching_number': False, 'is_account_reconcile': False, 'account_type': 'liability_current', 'account_internal_group': 'liability', 'account_root_id': (50053, '25'), 'display_type': 'tax', 'product_id': False, 'product_uom_id': False, 'product_uom_category_id': False, 'quantity': 0.0, 'date_maturity': False, 'price_unit': 0.0, 'price_subtotal': 0.0, 'price_total': 0.0, 'discount': 0.0, 'term_key': False, 'tax_key': {'tax_repartition_line_id': 702, 'group_tax_id': False, 'account_id': 1572, 'currency_id': 2, 'analytic_distribution': False, 'tax_ids': [(6, 0, [])], 'tax_tag_ids': [(6, 0, [])], 'partner_id': 124, 'move_id': 45, 'display_type': 'tax'}, 'compute_all_tax': {}, 'compute_all_tax_dirty': False, 'epd_key': False, 'epd_needed': False, 'epd_dirty': True, 'analytic_line_ids': [], 'discount_date': False, 'discount_amount_currency': 0.0, 'discount_balance': 0.0, 'discount_percentage': 0.0, 'blocked': False, 'is_refund': False, '__last_update':    datetime.datetime(2023, 4, 5, 8, 52, 6, 842107), 'display_name': 'INV/2019/00001 Tax 15%', 'create_uid': (24, 'Because I am accountman!'), 'create_date': datetime.datetime(2023, 4, 5, 8, 52, 6, 842107), 'write_uid': (24, 'Because I am accountman!'), 'write_date': datetime.datetime(2023, 4, 5, 8, 52, 6, 842107), 'move_attachment_ids': [], 'expected_pay_date': False, 'is_downpayment': False, 'sale_line_ids': [], 'asset_ids': [], 'non_deductible_tax_value': 0.0, 'followup_line_id': False, 'last_followup_date': False, 'next_action_date': False, 'invoice_date': datetime.date(2019, 1, 1), 'invoice_origin': False}]

</details>

william-andre reviewed (commented):
Didn't review fully yet, just quickly read.

The fix looks good otherwise :+1: (but I'd still like the concerns of @qdp-odoo to be clarified)

dutrieuc commented:
Hey @qdp-odoo if you have pointers on how you can reproduce the issue even with my fix that would be appreciated

qdp-odoo commented:
superseeded by https://github.com/odoo/odoo/pull/118443 in order to overcome the failed tests and be able to merge during cydu's absence
